ivy/functional/backends/torch/experimental/manipulation.py

```diff
@@ -328,3 +330,105 @@ def concat_from_sequence(
     elif new_axis == 1:
         ret = torch.stack(input_sequence, dim=axis)
         return ret
+
+
+def _can_handle_padding(input_shape, pad, mode):
+    n_dims = len(input_shape)
+    if n_dims == 0:
+        return False
+    if n_dims == 1 and mode != 'circular':
+        return False
+    if mode == 'circular':
+        return True
+    if isinstance(pad, int):
+        return False
+    for i in range(n_dims):
+        if not isinstance(pad[2 * i], int) or not isinstance(pad[2 * i + 1], int):
+            return False
+        if input_shape[i] + pad[2*i] + pad[2*i+1] < 0:
+            return False
+    return True
+
+
+def _check_tuple(t):
+    if isinstance(t, tuple):
+        if len(t) == 1 and not isinstance(t[0], tuple):
+            return t[0], True
+        else:
+            has_multiple_values = False
+            size = 0
+            for elem in t:
+                elem_val, elem_has_multiple_values = _check_tuple(elem)
+                if elem_has_multiple_values:
+                    has_multiple_values = True
+                size += elem_val
+            return size, has_multiple_values
+    else:
+        return 1, False
+
+
+def _check(*args, **kwargs):
+
+    mode = kwargs['mode']
+    if mode in ["linear_ramp",
+                    "maximum",
+                    "symmetric",
+                    "mean",
+                    "median",
+                    "minimum",
+                    "edge",
+                    "wrap",
+                    "empty",
+    ]:
+        return False
+    else:
+        inp = args[0]
+        pad = args[1]
+        if mode == 'constant':
+
+            c = kwargs['constant_values']
+            val, cond = _check_tuple(c)
+            if cond is False:
+                return False
+            else:
+                kwargs['constant_values'] = val
+                return True
+        elif _can_handle_padding(inp.shape, pad, mode):
+            return True
+        else:
```

Comment:
I don't think this is the correct approach for the `constant_values` and `pad`. You don't need to check these in the decorator to determine whether the backend will be used. You can convert them to the format that torch is expecting inside the backend function. This way we can use the native function for all the torch supported mode cases.

Reply:
If I apply this to the implementation of the native function then it does not work properly I think because these checks of tuples and dimensions return boolean True and False to represent the handler whether to call the compositional form or native form. so if we implement this into the TensorFlow or torch backend implementation then how it will call the compositional function at that time into the native implementation? on the other hand, we also need to use native implementation whenever it needs, so if we add this logic into the native implementation then it can be the same as compositional implementation because if we do this then it looks like we are refactoring these dimensions to meet the criteria I guess, so how would it differ from that?

Reply:
It would be more efficient to call a helper function to convert the padding and the constant value to the correct form inside the backend and then call the native function. If we just call the compositional implementation everything will be done manually, which is much slower.

Reply:
Torch only supports multi-dimensional padding forms in its backend, if we directly fetch the single dimension padding into torch, it raises an error on any mode. padding should be multi-dimensional in all the modes torch supports. it is really hard to convert the padding value into the correct dimension to call native implementation all time that's why I implemented this check helper function to check these padding forms and call the compositional when it torch backend does not support the constant and padding values. yes, it is less efficient than the native implementation because sometimes it calls compositional implementation but it also fulfills the work of the decorator to check whether to call the compositional or native implementation on some instructions. to check torch supports which kind of padding and constant values check this: [Click Here](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html)

Reply:
Could you give me an example of what you mean by single dimension vs multi dimension padding? I'm pretty sure it is possible as I've done conversions like this for the torch backend of the convolution functions.

Reply:
Also, please sync with master to resolve the conflicts.

Reply:
Torch's pad function expects the pad widths in a different format than numpy, jax and tensorflow, but they can all support all dimensionality cases.
Here is an example:
The input array is 3D, torch gets a 4D padding tuple and numpy gets a 3D tuple. The results are equivalent.
```
import torch
import torch.nn.functional as F
import numpy as np

t4d = torch.empty(3, 3, 4)
p3d = (2, 1, 3, 3)
out = F.pad(t4d, p3d, "constant", 0)
print(out.shape)
# torch.Size([3, 9, 7])

t = np.empty((3, 3, 4))
p = ((0, 0), (3, 3), (2, 1))
out = np.pad(t, p, 'constant', constant_values=0)
print(out.shape)
# (3, 9, 7)
```

Reply:
Hi @AnnaTz. Apologies for stepping in so late 😅. Although the conversation has gotten quite lengthy, if I followed correctly, I think what @usama-baloch is referring to w.r.t. the limited support for multidimensional padding is relevant for the other 3 modes of torch(i.e. `reflect`, `replicate`, and `circular`). *constant padding* is supported for arbitrary dimensions as is mentioned in the documentation [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html#:~:text=Constant%20padding%20is%20implemented%20for%20arbitrary%20dimensions.) which is why I think your example worked correctly. However, if the inputs are in a format that's not supported natively by torch, (for instance input is a 6-7 D tensor or the padding tuple is of odd length), in such cases, modifying the inputs to make it compatible with the native function whilst also guaranteeing identical results would become non-trivial. In such cases, it's better to resort to the compositional implementation, which is where the role of the `handle_mixed_functions` decorator comes in. 
  

@usama-baloch, with regards to the `_can_handle_padding` helper, it seems you're returning `True` when `mode=circular`. But shouldn't this be for `mode=constant` case? If I'm not mistaken, doesn't the `F.pad` function in torch impose some extra restrictions on the `pad` tuple and the `input` tensor for the other 3 modes so you cant always return `True` right?

Reply:
@YushaArif99 Yes, this is exactly what I'm trying to say.

@AnnaTz  @YushaArif99, with regards to the issue regarding the `_can_handle_padding` helper, I'm currently working on it and there are some cases that are failing.